File: streamlit_app/core/scoring.py
"""Logica di punteggio del fantamondiale 2026.

Regole:
  - 1X2 corretto: +2
  - ogni marcatore azzeccato: +1 (le doppiette contano, intersezione di multiset)
  - risultato esatto senza marcatori: +5 ; con almeno un marcatore: +10 (cumulativo)
  - se il 1X2 è sbagliato, TUTTO va a 0 (i marcatori si annullano).
"""

from __future__ import annotations

from dataclasses import dataclass

POINTS = {
    "outcome": 2,
    "per_scorer": 1,
    "exact_score": 5,
    "exact_score_with_scorer": 10,
}


def outcome(home: int, away: int) -> str:
    if home > away:
        return "1"
    if home < away:
        return "2"
    return "X"


def _scorer_key(s: dict) -> str:
    return f"{s['team']}|||{s['name']}"


def count_scorer_hits(predicted: list[dict], actual: list[dict]) -> int:
    """Numero di marcatori azzeccati, contando le doppiette."""
    need: dict[str, int] = {}
    for s in actual:
        k = _scorer_key(s)
        need[k] = need.get(k, 0) + 1
    hits = 0
    used: dict[str, int] = {}
    for s in predicted:
        k = _scorer_key(s)
        have = used.get(k, 0)
        if have < need.get(k, 0):
            hits += 1
            used[k] = have + 1
    return hits


@dataclass
class ScoreBreakdown:
    total: int
    correct_outcome: bool
    exact_score: bool
    scorer_hits: int
    points_outcome: int
    points_scorers: int
    points_exact_bonus: int
    voided: bool


def compute_score(pred: dict, result: dict) -> ScoreBreakdown:
    """pred/result: {home:int, away:int, scorers:[{team,name}]}."""
    correct = outcome(pred["home"], pred["away"]) == outcome(result["home"], result["away"])
    exact = pred["home"] == result["home"] and pred["away"] == result["away"]
    hits = count_scorer_hits(pred.get("scorers", []), result.get("scorers", []))

    if not correct:
        return ScoreBreakdown(
            total=0,
            correct_outcome=False,
            exact_score=False,
            scorer_hits=hits,
            points_outcome=0,
            points_scorers=0,
            points_exact_bonus=0,
            voided=hits > 0,
        )

    p_outcome = POINTS["outcome"]
    p_scorers = hits * POINTS["per_scorer"]
    if exact:
        p_exact = POINTS["exact_score_with_scorer"] if hits > 0 else POINTS["exact_score"]
    else:
        p_exact = 0

    return ScoreBreakdown(
        total=p_outcome + p_scorers + p_exact,
        correct_outcome=True,
        exact_score=exact,
        scorer_hits=hits,
        points_outcome=p_outcome,
        points_scorers=p_scorers,
        points_exact_bonus=p_exact,
        voided=False,
    )

File: src/lib/scoring.ts
import type { Prediction, MatchResult, ScorerRef } from "./types";

export const POINTS = {
  outcome: 2, // azzeccare 1X2
  perScorer: 1, // ogni marcatore azzeccato
  exactScore: 5, // risultato esatto (senza marcatori)
  exactScoreWithScorer: 10, // risultato esatto + almeno un marcatore
} as const;

export type Outcome = "1" | "X" | "2";

export function outcome(home: number, away: number): Outcome {
  if (home > away) return "1";
  if (home < away) return "2";
  return "X";
}

const scorerKey = (s: ScorerRef) => `${s.team}|||${s.name}`;

/** Numero di marcatori azzeccati, contando le doppiette (intersezione di multiset). */
export function countScorerHits(
  predicted: ScorerRef[],
  actual: ScorerRef[],
): number {
  const need = new Map<string, number>();
  for (const s of actual) need.set(scorerKey(s), (need.get(scorerKey(s)) ?? 0) + 1);
  let hits = 0;
  const used = new Map<string, number>();
  for (const s of predicted) {
    const k = scorerKey(s);
    const have = used.get(k) ?? 0;
    if (have < (need.get(k) ?? 0)) {
      hits++;
      used.set(k, have + 1);
    }
  }
  return hits;
}

export interface ScoreBreakdown {
  total: number;
  correctOutcome: boolean;
  exactScore: boolean;
  scorerHits: number;
  pointsOutcome: number;
  pointsScorers: number;
  pointsExactBonus: number;
  /** true quando i punti marcatori sono stati annullati per 1X2 sbagliato */
  voided: boolean;
}

/**
 * Regole:
 *  - 1X2 corretto: +2
 *  - ogni marcatore azzeccato: +1
 *  - risultato esatto senza marcatori: +5 ; con almeno un marcatore: +10
 *  - se il 1X2 è sbagliato, TUTTO va a 0 (i marcatori si annullano).
 */
export function computeScore(
  pred: Pick<Prediction, "home" | "away" | "scorers">,
  result: Pick<MatchResult, "home" | "away" | "scorers">,
): ScoreBreakdown {
  const correctOutcome = outcome(pred.home, pred.away) === outcome(result.home, result.away);
  const exactScore = pred.home === result.home && pred.away === result.away;
  const scorerHits = countScorerHits(pred.scorers, result.scorers);

  if (!correctOutcome) {
    return {
      total: 0,
      correctOutcome: false,
      exactScore: false,
      scorerHits,
      pointsOutcome: 0,
      pointsScorers: 0,
      pointsExactBonus: 0,
      voided: scorerHits > 0,
    };
  }

  const pointsOutcome = POINTS.outcome;
  const pointsScorers = scorerHits * POINTS.perScorer;
  const pointsExactBonus = exactScore
    ? scorerHits > 0
      ? POINTS.exactScoreWithScorer
      : POINTS.exactScore
    : 0;

  return {
    total: pointsOutcome + pointsScorers + pointsExactBonus,
    correctOutcome: true,
    exactScore,
    scorerHits,
    pointsOutcome,
    pointsScorers,
    pointsExactBonus,
    voided: false,
  };
}
